resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -283,7 +290,63 @@ private[spark] class KubernetesClusterSchedulerBackend(
     }
   }
 
-  private def allocateNewExecutorPod(): (String, Pod) = {
+  private def getCurrentNodesWithTaskCounts() : Map[String, Int] = {
+    val executorPodsWithIPs = EXECUTOR_PODS_BY_IPS_LOCK.synchronized {
+      executorPodsByIPs.values.toList  // toList makes a defensive copy.
+    }
+    val nodeToLocalTaskCount = mutable.Map[String, Int]() ++
+      KubernetesClusterSchedulerBackend.this.synchronized {
+        hostToLocalTaskCount
+      }
+    for (pod <- executorPodsWithIPs) {
+      // Remove cluster nodes that are running our executors already.
```

Comment:
Is preferring not-previously-used executors always a good idea?  It reminds me of the `spark.deploy.spreadOut` flag that Spark standalone has: https://spark.apache.org/docs/latest/spark-standalone.html

If you were doing a highly iterative workload that was network bound, it may be better to allow (or even prefer!) scheduling multiple executor pods on the same k8s node so that network is only on the loopback interface rather than actually crossing between servers.

Possibly we should consider adding a flag for this in the future.  The typical job I would run would prefer the spread out mode you have here so I'm not in a rush to support the consolidated mode.